src/aruku/walking/process/kinematic.cpp

```diff
@@ -0,0 +1,682 @@
+  m_a_move_amplitude_shift(0), m_arm_swing_gain(0), m_arm_roll_Gain(0), hip_comp(0.0), foot_comp(
+    0.0), time_unit(8.0), m_x_move_phase_shift(0.5_pi), m_x_move_amplitude_shift(0),
+  m_y_move_phase_shift(0.5_pi), m_z_move_phase_shift(0.5_pi), m_a_move_phase_shift(0.5_pi),
+  m_ctrl_running(false), m_real_running(false), m_time(0), x_move(0.0), y_move(0.0), a_move(0.0),
+  a_move_aim_on(false), is_compute_odometry(false)
+{
+  reset_angles();
+}
+
+void Kinematic::reset_angles()
+{
+  for (auto & angle : joint_angles) {
+    angle = 0.0;
+  }
+}
+
+const std::array<double, 18> & Kinematic::get_angles() const
+{
+  return joint_angles;
+}
+
+void Kinematic::set_running_state(bool running_state)
+{
+  m_ctrl_running = running_state;
+
+  if (running_state) {
+    m_real_running = running_state;
+  }
+}
+
+bool Kinematic::get_running_state() const
+{
+  return m_ctrl_running;
+}
+
+void Kinematic::set_move_amplitude(double x, double y, double a, bool aim_on)
+{
+  x_move = x;
+  y_move = y;
+  a_move = a;
+  a_move_aim_on = aim_on;
+}
+
+double Kinematic::get_x_move_amplitude() const
+{
+  return m_x_move_amplitude;
+}
+
+double Kinematic::get_y_move_amplitude() const
+{
+  return m_y_move_amplitude;
+}
+
+double Kinematic::get_hip_offest() const
+{
+  return hip_pitch_offset + hip_comp;
+}
+
+bool Kinematic::time_to_compute_odometry() const
+{
+  return is_compute_odometry;
+}
+
+void Kinematic::stop_kinematic()
+{
+  m_ctrl_running = false;
+  m_real_running = false;
+
+  m_x_move_amplitude = 0.0;
+  m_y_move_amplitude = 0.0;
+  m_a_move_amplitude = 0.0;
+  m_z_move_amplitude = 0.0;
+
+  update_times();
+  update_move_amplitude();
+}
+
+double Kinematic::wsin(
+  double time, double period, double period_shift, double mag,
+  double mag_shift) const
+{
+  return mag * sin(2_pi / period * time - period_shift) + mag_shift;
+}
+
+bool Kinematic::compute_inverse_kinematic(
+  std::string leg, double x, double y, double z, double a,
+  double b, double c)
+{
+  size_t index = (leg == "right") ? 0 : 6;
+
+  {
+    using keisan::Point3;
+    using keisan::Euler;
+    using keisan::Matrix;
+
+    auto matrix_translation = keisan::translation_matrix(Point3(x, y, z - leg_length));
+    auto matrix_rotation = keisan::rotation_matrix(
+      Euler(
+        keisan::make_radian(a),
+        keisan::make_radian(b), keisan::make_radian(c)));
+
+    auto matrix_transformation = matrix_translation * matrix_rotation;
+
+    auto vector = Point3(
+      x + (matrix_transformation[0][2] * ankle_length),
+      y + (matrix_transformation[1][2] * ankle_length),
+      (z - leg_length) + (matrix_transformation[2][2] * ankle_length));
+
+    // get knee
+    double vector_magnitude = vector.magnitude();
+    double acos_result = acos(
+      ((vector_magnitude * vector_magnitude) - (thigh_length * thigh_length) -
+      (calf_length * calf_length)) / (2 * thigh_length * calf_length));
```

Comment:
can we later add this function in keisan? seems this equation is often being used in this function 😆 

angle of triangle equation
<img width="170" alt="image" src="https://user-images.githubusercontent.com/10202888/158953786-7caa9f69-9891-4a29-bec6-a996e86fbc70.png">

Reply:
noted it